### PR TITLE
arch: arm: update compile options for DSP

### PR DESCRIPTION
**Pull request by ioannisg**
This commit updates the compile options in cmake to distinguish
between ARMv8-M Mainline MCUs with or withouth the optional DSP
extension. A new Kconfig option: ARMV8_M_DSP, is introduced, to
signify the use of an ARMv8-M MCU with DSP support.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

### CODE REVIEW

File: arch/arm/CMakeLists.txt [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
-set(ARCH_FOR_cortex-m3     armv7-m     )
-set(ARCH_FOR_cortex-m4     armv7e-m    )
-set(ARCH_FOR_cortex-m23    armv8-m.base)
-set(ARCH_FOR_cortex-m33    armv8-m.main)
+set(ARCH_FOR_cortex-m0        armv6-m         )
+set(ARCH_FOR_cortex-m0plus    armv6-m         )
+set(ARCH_FOR_cortex-m3        armv7-m         )
+set(ARCH_FOR_cortex-m4        armv7e-m        )
+set(ARCH_FOR_cortex-m23       armv8-m.base    )
+set(ARCH_FOR_cortex-m33       armv8-m.main+dsp)
```
Comment (carlescufi) [inline arch/arm/CMakeLists.txt]: Why not do it the other way around? 
```
ARCH_FOR_cortex-m33       armv8-m.main
ARCH_FOR_cortex-m33+dsp armv8-m.main+dsp
```

Comment (ioannisg) [reply arch/arm/CMakeLists.txt]: @carlescufi mcpu=cortex_m33 is _incompatible_ with march=armv8-m.main; GCC compiler gives a warning:

>cc1.exe: warning: switch -mcpu=cortex-m33 conflicts with -march=armv8-m.main switch

Comment (carlescufi) [reply arch/arm/CMakeLists.txt]: ok understood

**Review by carlescufi — APPROVED** [review]

File: arch/arm/CMakeLists.txt (hunk 1 above)
Comment (galak) [reply arch/arm/CMakeLists.txt]: Is this related to the bug we the current ARM toolchain?

Comment (ioannisg) [reply arch/arm/CMakeLists.txt]: @galak I am not sure about it.
AFAIK, the "bug" is the conflict with the SVC instruction support (related PR; #7445)

GCC tool-chain does not support the -march=armv8-m.main+nodsp option yet. However,
as of its latest release (2008 7 update 2, see Release notes) the tool-chain has started supporting the -mcpu=cortex-m33+nodsp option.

This makes me think that the tool-chain assumes, now, the following:
- cortex-m33+nodsp compatible with: armv8-m.main
- cortex-m33 compatible with armv8-m.main+dsp

 This options allow to build with or without the DSP support, by solely differentiating on the -mcpu option.

**Review by tyrantlizard77 — APPROVED** [review]

File: arch/arm/CMakeLists.txt (hunk 1 above)
Comment (galak) [reply arch/arm/CMakeLists.txt]: I think there was the issue that the 2017 gcc release didn't support cortex-m33+nodsp, and that cortex-m33 implied the dsp options.

**Review by galak — APPROVED** [review]

**Pull request merged**